### PR TITLE
HHH-18628 Regression: Unable to determine TableReference

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-18628

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/CriteriaSelectJoinTest.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+@Jpa(
+		annotatedClasses = {
+				CriteriaSelectJoinTest.Parent.class,
+				CriteriaSelectJoinTest.Child.class,
+		}
+)
+@JiraKey( "HHH-18628" )
+public class CriteriaSelectJoinTest {
```
Comment (mbladel) [inline hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/CriteriaSelectJoinTest.java:36]: Small nitpick, maybe `CriteriaSelectOneToOneUnownedTest` would be more specific to what this is doing.

File: hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/CriteriaSelectJoinTest.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+				entityManager -> {
+					CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+
+					CriteriaQuery<Child> query = cb.createQuery( Child.class );
+					Root<Parent> parent = query.from( Parent.class );
+					Join<Parent, Child> child = parent.join( "child", JoinType.INNER );
+					query.select( child );
+
+					List<Child> children = entityManager.createQuery( query ).getResultList();
+					assertThat( children ).isNotNull();
```
Comment (mbladel) [inline hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/CriteriaSelectJoinTest.java:72]: Would be nice to also test a `JoinType.LEFT` version of the query, just to ensure we preserve the correct behavior in all cases.

Also, maybe adding a more meaningful assertion on the actual data returned by the query would add value.

**Review by mbladel — COMMENTED** [review]
LGTM, I think the check was forgotten while refactoring but let's hear from @beikov maybe there was a reason for it. Left a couple minor nitpicks.

File: hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/CriteriaSelectJoinTest.java (hunk 2 above)
Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/orm/test/jpa/criteria/CriteriaSelectJoinTest.java]: Thanks @mbladel applied your suggestions.

**Review by beikov — APPROVED** [review]
I tested the impact of the change on various queries and it all works fine. Thanks!

**Comment by dreab8** [comment]
Thanks @mbladel @beikov

**Pull request merged**